compiler/codegen/codegen.cpp

```diff
@@ -2071,11 +2071,22 @@ void codegen(void) {
   }
 
   // Set the executable name to the name of the file containing the
-  // main module (minus its extension) if it isn't set already.
+  // main module (minus its path and extension) if it isn't set
+  // already.
   if (executableFilename[0] == '\0') {
     ModuleSymbol* mainMod = ModuleSymbol::mainModule();
-    strncpy(executableFilename, mainMod->astloc.filename,
-            sizeof(executableFilename));
+    const char* mainModFilename = mainMod->astloc.filename;
+
+    // find the last slash in the filename's path, if there is one
+    const char* lastSlash = strrchr(mainModFilename, '/');
+    if (lastSlash == NULL) {
+      lastSlash = mainModFilename;
+    } else {
+      lastSlash++;
+    }
+
+    // copy from that slash onwards into the executableFilename
+    strncpy(executableFilename, lastSlash, sizeof(executableFilename));
```

Comment:
This is an opportunity to fix a subtle bug that existed in the old code prior to this change.  The `FILENAME_MAX` macro that is used to set the size of the `executableFilename` array is not guaranteed to be the longest filename on the system.  It is only guaranteed to be the longest filename length that can _always_ be used.

That means that `yyfilename`, which is used to set `astloc.filename`, could conceivably be longer than `FILENAME_MAX`.

Now, `strncpy()` will yield a non-null-terminated string if its input string is at least as large as its output buffer.  That will lead to a buffer overrun later.

To fix this, we need to copy in `sizeof(executableFilename)-1` bytes and explicitly set `executableFilename[sizeof(executableFilename)-1]` to `'\0'`.  (An alternative would be to make `executableFilename` dynamically allocated, but we don't have to do that.)

Reply:
:+1:

Though it also seems surprising to produce an executable that's a truncated version of the source filename.

Reply:
I agree with Paul that it seems odd to quietly truncate the executable filename in the event that this overrun does occur.  What if we were to throw an internal error if `strlen(executableFilename)` exceeded `FILENAME_MAX`? (given how unlikely it is to occur for normal use cases)